lib/src/data_structures.rs

```diff
@@ -18,6 +18,863 @@ use std::slice::{Iter, IterMut};
+// RegSet
+
+const BLOCK_SIZE: usize = 64;
+
+pub struct BitSet {
+  data: Vec<u64>,
+}
+
+pub trait RegSet {
+  fn empty() -> Self;
+  /// Return a new BitSet with one item set.
+  fn unit(item: Reg) -> Self;
+  /// Return a new BitSet with two items set.
+  fn two(item1: Reg, item2: Reg) -> Self;
+  fn card(&self) -> usize;
+  /// Set the item to 1.
+  fn insert(&mut self, item: Reg);
+  /// If the item is set, clear the item to 0.
+  /// Do nothing otherwise.
+  fn delete(&mut self, item: Reg);
+  fn is_empty(&self) -> bool;
+  /// Toggle the item.
+  fn contains(&self, item: Reg) -> bool;
+  fn intersect(&mut self, other: &Self);
+  fn union(&mut self, other: &Self);
+  fn remove(&mut self, other: &Self);
+  fn intersects(&self, other: &Self) -> bool;
+  fn is_subset_of(&self, other: &Self) -> bool;
+  fn equals(&self, other: &Self) -> bool;
+}
+
+impl RegSet for BitSet {
+  fn empty() -> Self {
+    let mut d = Vec::<u64>::new();
+    d.push(0);
+    Self { data: d }
+  }
+
+  fn unit(item: Reg) -> Self {
+    let mut s: BitSet = RegSet::empty();
+    s.insert(item);
+    s
+  }
+
+  fn two(item1: Reg, item2: Reg) -> Self {
+    let mut s: BitSet = RegSet::empty();
+    s.insert(item1);
+    s.insert(item2);
+    s
+  }
+
+  fn card(&self) -> usize {
+    let mut counter = 0;
+
+    for i in 0..self.data.len() {
+      counter += self.data[i].count_ones();
+    }
+    counter as usize
+  }
+
+  fn insert(&mut self, item: Reg) {
+    let reg_index = item.get_index();
+    let data_index = reg_index / BLOCK_SIZE;
+    let offset = reg_index % BLOCK_SIZE;
+
+    while data_index >= self.data.len() {
+      self.data.push(0);
+    }
+    self.data[data_index] |= 1 << offset;
+  }
+
+  fn delete(&mut self, item: Reg) {
+    let reg_index = item.get_index();
+    let data_index = reg_index / BLOCK_SIZE;
+    let offset = reg_index % BLOCK_SIZE;
+
+    self.data[data_index] &= !(1 << offset);
+  }
+
+  fn is_empty(&self) -> bool {
+    for i in 0..self.data.len() {
+      if self.data[i] != 0 {
+        return false;
+      }
+    }
+    true
+  }
+
+  fn contains(&self, item: Reg) -> bool {
+    let reg_index = item.get_index();
+    let data_index = reg_index / BLOCK_SIZE;
+    let offset = reg_index % BLOCK_SIZE;
+
+    if data_index >= self.data.len() {
+      false
+    } else {
+      (1 & (self.data[data_index] >> offset)) != 0
+    }
+  }
+
+  fn intersect(&mut self, other: &Self) {
+    let vec_size = {
+      if other.data.len() < self.data.len() {
+        other.data.len()
+      } else {
+        self.data.len()
+      }
+    };
+
+    for i in 0..vec_size {
+      self.data[i] &= other.data[i];
+    }
+    for i in vec_size..self.data.len() {
+      self.data[i] = 0;
+    }
+  }
+
+  fn union(&mut self, other: &Self) {
+    let vec_size = {
+      if other.data.len() <= self.data.len() {
+        other.data.len()
+      } else {
+        for _ in 0..(other.data.len() - self.data.len()) {
```

Comment:
Maybe more natural as

```suggestion
        for _ in self.data.len()..other.data.len() {
```